File: mobile-command-bridge/android-client/termux-client.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");
const { WebSocket } = require("ws");
const readline = require("readline");
const crypto = require("crypto");

const HISTORY_FILE = path.resolve(__dirname, "command-history.json");
const MAX_HISTORY_ITEMS = 100;

const PRESET_TASKS = {
  prepare: ["task.run", "browser.request.prepare"],
  watch: ["task.run", "browser.request.watch"],
  complete: ["task.run", "browser.request.complete"]
};

const PRESET_VSCODE = {
  chat: ["vscode.command", "workbench.action.chat.open"],
  ask: ["vscode.command", "workbench.action.chat.openAsk"],
  copilot: ["vscode.command", "github.copilot.chat.open"]
};

function parseArgs() {
  const args = process.argv.slice(2);
  const out = {
    host: "127.0.0.1",
    port: 8787,
    token: "",
    onceAction: "",
    onceTarget: "",
    oncePrompt: ""
  };

  for (let i = 0; i < args.length; i += 1) {
    const arg = args[i];
    if (arg === "--host") {
      out.host = args[i + 1];
      i += 1;
      continue;
    }

    if (arg === "--port") {
      out.port = Number(args[i + 1]);
      i += 1;
      continue;
    }

    if (arg === "--token") {
      out.token = args[i + 1];
      i += 1;
      continue;
    }

    if (arg === "--once") {
      out.onceAction = args[i + 1] || "";
      out.onceTarget = args[i + 2] || "";
      i += 2;
      continue;
    }

    if (arg === "--once-prompt") {
      out.oncePrompt = args[i + 1] || "";
      i += 1;
      continue;
    }
  }

  if (!out.token) {
    console.error("Missing --token. Example: node termux-client.js --host 192.168.1.42 --token <BRIDGE_TOKEN>");
    process.exit(1);
  }

  return out;
}

function generateRequestId() {
  return `req-${Date.now()}-${crypto.randomBytes(3).toString("hex")}`;
}

function generateNonce() {
  return `nonce-${crypto.randomBytes(8).toString("hex")}`;
}

function printHelp() {
  console.log("\nCommands:");
  console.log("  help                              Show this help");
  console.log("  run <prepare|watch|complete>      Run task preset");
  console.log("  vc <chat|ask|copilot>             Run VS Code command preset");
  console.log("  exec <action> <target>            Run custom allowlisted action/target");
  console.log("  agent <name> <text...>            Dispatch prompt to agent (AutoHotkey path)");
  console.log("  draft <name> <text...>            Write prompt artifact without dispatch");
  console.log("  history [count]                   Show recent execute command history");
  console.log("  retry <index>                     Retry a history entry by index from history output");
  console.log("  last                              Retry most recent history entry");
  console.log("  cancel [requestId]                Cancel last request or specific requestId");
  console.log("  exit                              Quit");
}

function loadHistory() {
  if (!fs.existsSync(HISTORY_FILE)) {
    return [];
  }

  try {
    const raw = fs.readFileSync(HISTORY_FILE, "utf8");
    const parsed = JSON.parse(raw);
    if (!Array.isArray(parsed)) {
      return [];
    }

    return parsed.filter((item) => isHistoryEntry(item));
  } catch {
    return [];
  }
}

function saveHistory(history) {
  const trimmed = history.slice(0, MAX_HISTORY_ITEMS);
  fs.writeFileSync(HISTORY_FILE, JSON.stringify(trimmed, null, 2), "utf8");
}

function isHistoryEntry(item) {
  if (!item || typeof item !== "object") {
    return false;
  }

  if (typeof item.action !== "string" || typeof item.target !== "string") {
    return false;
  }

  if (item.promptText !== undefined && typeof item.promptText !== "string") {
    return false;
  }

  return true;
}

function appendHistory(history, entry) {
  const normalized = {
    createdAt: new Date().toISOString(),
    action: entry.action,
    target: entry.target,
    promptText: entry.promptText || ""
  };

  history.unshift(normalized);
  saveHistory(history);
}

function printHistory(history, countInput) {
  const requestedCount = Number(countInput || 10);
  const count = Number.isFinite(requestedCount) && requestedCount > 0 ? requestedCount : 10;
  const shown = history.slice(0, count);

  if (shown.length === 0) {
    console.log("No history entries yet.");
    return;
  }

  console.log("Recent execute history:");
  for (let i = 0; i < shown.length; i += 1) {
    const item = shown[i];
    const suffix = item.promptText ? ` prompt=\"${item.promptText.slice(0, 80)}${item.promptText.length > 80 ? "..." : ""}\"` : "";
    console.log(`${i + 1}. ${item.action} ${item.target}${suffix} @ ${item.createdAt}`);
  }
}

function getHistoryEntry(history, oneBasedIndex) {
  const idx = Number(oneBasedIndex);
  if (!Number.isFinite(idx) || idx <= 0) {
    return null;
  }

  return history[idx - 1] || null;
}

function connect() {
  const opts = parseArgs();
  const url = `ws://${opts.host}:${opts.port}/ws`;
  const interactiveMode = !(opts.onceAction && opts.onceTarget);
  const history = loadHistory();
  let lastRequestId = null;
  let oneShotRequestId = null;
  const pendingPayloads = [];

  const ws = new WebSocket(url, {
    headers: {
      "x-bridge-token": opts.token
    }
  });

  ws.on("open", () => {
    console.log(`Connected to ${url}`);
    if (pendingPayloads.length > 0) {
      console.log(`Flushing ${pendingPayloads.length} queued command(s)`);
      for (const payload of pendingPayloads.splice(0)) {
        ws.send(JSON.stringify(payload));
        console.log(`> ${payload.type} ${payload.requestId || ""} ${payload.action || ""} ${payload.target || ""}`);
      }
    }

    if (opts.onceAction && opts.onceTarget) {
      const extra = opts.oncePrompt ? { promptText: opts.oncePrompt } : {};
      sendExecute(opts.onceAction, opts.onceTarget, extra);
      return;
    }

    printHelp();
    rl.prompt();
  });

  ws.on("message", (data) => {
    let event;
    try {
      event = JSON.parse(data.toString("utf8"));
    } catch {
      console.log(`< raw ${data.toString("utf8")}`);
      rl.prompt();
      return;
    }

    const type = event.type;
    if (type === "stream") {
      const chunk = event.chunk || "";
      process.stdout.write(`< [${event.requestId}] ${event.stream}: ${chunk}`);
      if (!chunk.endsWith("\n")) {
        process.stdout.write("\n");
      }
      if (interactiveMode) {
        rl.prompt();
      }
      return;
    }

    if (type === "ready") {
      console.log(`< ready version=${event.version}`);
      if (interactiveMode) {
        rl.prompt();
      }
      return;
    }

    if (type === "started") {
      console.log(`< started ${event.requestId} ${event.action} ${JSON.stringify(event.meta || {})}`);
      if (interactiveMode) {
        rl.prompt();
      }
      return;
    }

    if (type === "complete") {
      console.log(`< complete ${event.requestId} exitCode=${event.exitCode} signal=${event.signal}`);
      if (oneShotRequestId && event.requestId === oneShotRequestId) {
        ws.close();
        return;
      }
      if (interactiveMode) {
        rl.prompt();
      }
      return;
    }

    if (type === "error") {
      console.log(`< error requestId=${event.requestId || "-"} code=${event.code} message=${event.message}`);
      if (oneShotRequestId && event.requestId === oneShotRequestId) {
        ws.close();
        return;
      }
      if (interactiveMode) {
        rl.prompt();
      }
      return;
    }

    if (type === "cancelled") {
      console.log(`< cancelled ${event.requestId} cancelled=${event.cancelled}`);
      if (interactiveMode) {
        rl.prompt();
      }
      return;
    }

    console.log(`< ${JSON.stringify(event)}`);
    if (interactiveMode) {
      rl.prompt();
    }
  });

  ws.on("close", (code, reason) => {
    console.log(`Disconnected (${code}): ${String(reason || "")}`);
    process.exit(0);
  });

  ws.on("error", (error) => {
    console.error(`WebSocket error: ${error.message}`);
  });

  function send(payload) {
    if (ws.readyState !== WebSocket.OPEN) {
      pendingPayloads.push(payload);
      console.log(`(queued) ${payload.type} ${payload.requestId || ""} ${payload.action || ""} ${payload.target || ""}`);
      return;
    }

    ws.send(JSON.stringify(payload));
    console.log(`> ${payload.type} ${payload.requestId || ""} ${payload.action || ""} ${payload.target || ""}`);
  }

  function sendExecute(action, target, extra = {}) {
    const payload = {
      type: "execute",
      requestId: generateRequestId(),
      action,
      target,
      timestamp: Date.now(),
      nonce: generateNonce(),
      ...extra
    };
    lastRequestId = payload.requestId;
    if (opts.onceAction && opts.onceTarget) {
      oneShotRequestId = payload.requestId;
    }
    appendHistory(history, {
      action,
      target,
      promptText: typeof extra.promptText === "string" ? extra.promptText : ""
    });
    send(payload);
  }

  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
    prompt: "bridge> "
  });

  rl.on("line", (line) => {
    const raw = line.trim();
    if (!raw) {
      rl.prompt();
      return;
    }

    const parts = raw.split(/\s+/);
    const cmd = parts[0].toLowerCase();

    if (cmd === "help") {
      printHelp();
      rl.prompt();
      return;
    }

    if (cmd === "exit") {
      rl.close();
      ws.close();
      return;
    }

    if (cmd === "run" && parts[1]) {
      const preset = PRESET_TASKS[parts[1].toLowerCase()];
      if (!preset) {
        console.log("Unknown task preset");
        rl.prompt();
        return;
      }
      sendExecute(preset[0], preset[1]);
      rl.prompt();
      return;
    }

    if (cmd === "vc" && parts[1]) {
      const preset = PRESET_VSCODE[parts[1].toLowerCase()];
      if (!preset) {
        console.log("Unknown VS Code preset");
        rl.prompt();
        return;
      }
      sendExecute(preset[0], preset[1]);
      rl.prompt();
      return;
    }

    if (cmd === "exec" && parts.length === 3) {
      sendExecute(parts[1], parts[2]);
      rl.prompt();
      return;
    }

    if (cmd === "agent" && parts.length >= 3) {
      const target = parts[1].toLowerCase();
      const promptText = parts.slice(2).join(" ");
      sendExecute("prompt.dispatch", target, { promptText });
      rl.prompt();
      return;
    }

    if (cmd === "draft" && parts.length >= 3) {
      const target = parts[1].toLowerCase();
      const promptText = parts.slice(2).join(" ");
      sendExecute("prompt.write", target, { promptText });
      rl.prompt();
      return;
    }

    if (cmd === "cancel") {
      const requestId = parts[1] || lastRequestId;
      if (!requestId) {
        console.log("No requestId available to cancel");
        rl.prompt();
        return;
      }
      send({
        type: "cancel",
        requestId
      });
      rl.prompt();
      return;
    }

    if (cmd === "history") {
      printHistory(history, parts[1]);
      rl.prompt();
      return;
    }

    if (cmd === "retry") {
      const entry = getHistoryEntry(history, parts[1]);
      if (!entry) {
        console.log("Invalid history index. Use: history");
        rl.prompt();
        return;
      }
      const extra = entry.promptText ? { promptText: entry.promptText } : {};
      sendExecute(entry.action, entry.target, extra);
      rl.prompt();
      return;
    }

    if (cmd === "last") {
      const entry = history[0];
      if (!entry) {
        console.log("No history entries yet.");
        rl.prompt();
        return;
      }
      const extra = entry.promptText ? { promptText: entry.promptText } : {};
      sendExecute(entry.action, entry.target, extra);
      rl.prompt();
      return;
    }

    console.log("Unknown command. Type 'help'.");
    rl.prompt();
  });
}

connect();

File: mobile-command-bridge/android-client/termux-client.py
#!/usr/bin/env python3
"""Simple interactive Android/Termux client for the mobile command bridge."""

import argparse
import json
import queue
import secrets
import threading
import time
from dataclasses import dataclass

import websocket


PRESET_TASKS = {
    "prepare": ("task.run", "browser.request.prepare"),
    "watch": ("task.run", "browser.request.watch"),
    "complete": ("task.run", "browser.request.complete"),
}

PRESET_VSCODE = {
    "chat": ("vscode.command", "workbench.action.chat.open"),
    "ask": ("vscode.command", "workbench.action.chat.openAsk"),
    "copilot": ("vscode.command", "github.copilot.chat.open"),
}


@dataclass
class SessionState:
    ws: websocket.WebSocketApp | None = None
    ready: bool = False
    last_request_id: str | None = None


def generate_request_id() -> str:
    return f"req-{int(time.time() * 1000)}-{secrets.token_hex(3)}"


def generate_nonce() -> str:
    return f"nonce-{secrets.token_hex(8)}"


def make_execute_payload(action: str, target: str) -> dict:
    request_id = generate_request_id()
    return {
        "type": "execute",
        "requestId": request_id,
        "action": action,
        "target": target,
        "timestamp": int(time.time() * 1000),
        "nonce": generate_nonce(),
    }


def make_cancel_payload(request_id: str) -> dict:
    return {
        "type": "cancel",
        "requestId": request_id,
    }


def print_help() -> None:
    print("\nCommands:")
    print("  help                              Show this help")
    print("  run <prepare|watch|complete>      Run task preset")
    print("  vc <chat|ask|copilot>             Run VS Code command preset")
    print("  exec <action> <target>            Run a custom allowlisted action/target")
    print("  cancel [requestId]                Cancel last request or a specific one")
    print("  exit                              Quit")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Termux client for mobile-command-bridge")
    parser.add_argument("--host", default="127.0.0.1", help="Laptop host/IP")
    parser.add_argument("--port", type=int, default=8787, help="Bridge port")
    parser.add_argument("--token", required=True, help="Bridge token (BRIDGE_TOKEN)")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    state = SessionState()
    event_queue: queue.Queue[dict] = queue.Queue()

    ws_url = f"ws://{args.host}:{args.port}/ws"

    def on_open(_: websocket.WebSocketApp) -> None:
        print(f"Connected to {ws_url}")

    def on_message(_: websocket.WebSocketApp, message: str) -> None:
        try:
            payload = json.loads(message)
        except json.JSONDecodeError:
            print(f"[raw] {message}")
            return

        event_queue.put(payload)

    def on_error(_: websocket.WebSocketApp, error: Exception) -> None:
        print(f"[error] websocket: {error}")

    def on_close(_: websocket.WebSocketApp, status_code: int, close_msg: str) -> None:
        print(f"Disconnected ({status_code}): {close_msg}")

    headers = [f"x-bridge-token: {args.token}"]

    ws = websocket.WebSocketApp(
        ws_url,
        header=headers,
        on_open=on_open,
        on_message=on_message,
        on_error=on_error,
        on_close=on_close,
    )
    state.ws = ws

    thread = threading.Thread(target=ws.run_forever, daemon=True)
    thread.start()

    # Wait briefly for the initial ready event.
    ready_deadline = time.time() + 5
    while time.time() < ready_deadline and not state.ready:
        pump_events(state, event_queue)
        time.sleep(0.1)

    print_help()

    while True:
        pump_events(state, event_queue)
        try:
            raw = input("bridge> ").strip()
        except (EOFError, KeyboardInterrupt):
            print("\nBye")
            break

        if not raw:
            continue

        parts = raw.split()
        cmd = parts[0].lower()

        if cmd == "help":
            print_help()
            continue

        if cmd == "exit":
            break

        if cmd == "run" and len(parts) == 2:
            preset = PRESET_TASKS.get(parts[1].lower())
            if not preset:
                print("Unknown task preset")
                continue
            send_execute(state, *preset)
            continue

        if cmd == "vc" and len(parts) == 2:
            preset = PRESET_VSCODE.get(parts[1].lower())
            if not preset:
                print("Unknown VS Code preset")
                continue
            send_execute(state, *preset)
            continue

        if cmd == "exec" and len(parts) == 3:
            send_execute(state, parts[1], parts[2])
            continue

        if cmd == "cancel":
            request_id = parts[1] if len(parts) == 2 else state.last_request_id
            if not request_id:
                print("No requestId available to cancel")
                continue
            send_payload(state, make_cancel_payload(request_id))
            continue

        print("Unknown command. Type 'help'.")

    if state.ws:
        state.ws.close()


def send_execute(state: SessionState, action: str, target: str) -> None:
    payload = make_execute_payload(action, target)
    state.last_request_id = payload["requestId"]
    send_payload(state, payload)


def send_payload(state: SessionState, payload: dict) -> None:
    if not state.ws:
        print("Not connected")
        return

    state.ws.send(json.dumps(payload))
    print(f"-> {payload['type']} {payload.get('requestId', '')} {payload.get('action', '')} {payload.get('target', '')}")


def pump_events(state: SessionState, event_queue: queue.Queue[dict]) -> None:
    while True:
        try:
            event = event_queue.get_nowait()
        except queue.Empty:
            return

        event_type = event.get("type", "unknown")

        if event_type == "ready":
            state.ready = True
            print(f"<- ready version={event.get('version')} serverTime={event.get('serverTime')}")
            continue

        if event_type == "stream":
            req = event.get("requestId", "?")
            stream_name = event.get("stream", "stdout")
            chunk = event.get("chunk", "")
            print(f"<- [{req}] {stream_name}: {chunk}", end="" if chunk.endswith("\n") else "\n")
            continue

        if event_type == "started":
            print(f"<- started {event.get('requestId')} {event.get('action')} {event.get('meta', {})}")
            continue

        if event_type == "complete":
            print(
                f"<- complete {event.get('requestId')} exitCode={event.get('exitCode')} signal={event.get('signal')}"
            )
            continue

        if event_type == "cancelled":
            print(f"<- cancelled {event.get('requestId')} cancelled={event.get('cancelled')}")
            continue

        if event_type == "error":
            print(
                f"<- error requestId={event.get('requestId')} code={event.get('code')} message={event.get('message')}"
            )
            continue

        print(f"<- {json.dumps(event)}")


if __name__ == "__main__":
    main()
